lib/Conversion/LWEToPolynomial/LWEToPolynomial.cpp

```diff
@@ -0,0 +1,126 @@
+#include "lib/Conversion/LWEToPolynomial/LWEToPolynomial.h"
+
+#include <cstddef>
+#include <optional>
+#include <utility>
+
+#include "lib/Conversion/Utils.h"
+#include "lib/Dialect/LWE/IR/LWEOps.h"
+#include "lib/Dialect/LWE/IR/LWETypes.h"
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Polynomial/IR/PolynomialOps.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Polynomial/IR/PolynomialTypes.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/BuiltinOps.h"             // from @llvm-project
+#include "mlir/include/mlir/IR/BuiltinTypes.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/ValueRange.h"             // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir::heir::lwe {
+
+#define GEN_PASS_DEF_LWETOPOLYNOMIAL
+#include "lib/Conversion/LWEToPolynomial/LWEToPolynomial.h.inc"
+
+class CiphertextTypeConverter : public TypeConverter {
+ public:
+  // Convert ciphertext to tensor<#dim x !poly.poly<#rings[#level]>>
+  // Our precondition (not verified) is that the key and plaintext have only one
+  // dimension, and the ciphertext has two dimensions. (i.e.) The key is a
+  // single polynomial and not a tensor of higher dimensions. We may support
+  // higher dimensions in the future (for schemes such as TFHE). Here are some
+  // TODO:
+  // 1. Add an encrypt/decrypt verifier to check for the preconditions of
+  // encrypt and decrypt. See lib/Target/Verilog/VerilgEmitter.cpp for an
+  // example of op->walk to use in verifying.
+  // 2. For TFHE, which can support higher dimensional keys, plaintexts, and
+  // ciphertexts, we need to add support for encrypt and decrypt for
+  // those cases.
+  CiphertextTypeConverter(MLIRContext *ctx) {
```

Comment:
This converter (at least, the ctxt conversion) also appears in `--bgv-to-polynomial`, so I was going to suggest refactoring this so that both passes use the same converter.
However, we could also split `--bgv-to-polynomial` into `--bgv-to-lwe` and `--lwe-to-polynomial`. In my draft `bgv.relinearize` lowering, I'm already emitting mostly `lwe` ops and few actual `polynomial` ops, so `--bgv-to-polynomial` wasn't really accurate anymore anyway.

Making `lwe` a full abstraction layer (e.g, adding `lwe.add/sub` and `lwe.tensor_style_mul`) would be nice as it'd allow us to reuse those lowerings for CKKS, too, rather than repeating them across BGV/BFV and CKKS. Especially if we want to support degree > 2 in mult and relins, we'd be repeating a good chunk of code each time.

Reply:
+1 we probably want to have this `--lwe-to-polynomial` be a common pass that is applied within BGV -> standard, CKKS -> standard, and even TFHE -> standard.

Adding!

Minimally for this PR we can move the type converter here into a separate file (LweToPolynomial/TypeConverter.cpp), have both passes use it for now, and then in a future PR split and refactor the `--bgv-to-polynomial` into `--bgv-to-lwe` so that it doesn't use the type converter anymore. But let's chat tomorrow if that's the direction we want to go

Reply:
Filing a follow-up https://github.com/google/heir/issues/828